File: examples/payload-hash/python/mail_payload_hash.py
"""Mailer payload_hash helper for Python consumers."""

from __future__ import annotations

import hashlib
import json
from typing import Any

INCLUDED_FIELDS = frozenset(
    [
        "source_service",
        "purpose",
        "to",
        "subject",
        "html_body",
        "text_body",
        "reply_to",
        "metadata",
    ]
)


def escape_json_string(value: str) -> str:
    parts = ['"']
    for character in value:
        code = ord(character)
        if character == '"':
            parts.append('\\"')
        elif character == "\\":
            parts.append("\\\\")
        elif character == "\b":
            parts.append("\\b")
        elif character == "\f":
            parts.append("\\f")
        elif character == "\n":
            parts.append("\\n")
        elif character == "\r":
            parts.append("\\r")
        elif character == "\t":
            parts.append("\\t")
        elif code < 0x20:
            parts.append(f"\\u{code:04x}")
        else:
            parts.append(character)
    parts.append('"')
    return "".join(parts)


def canonicalize(value: Any) -> str:
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, str):
        return escape_json_string(value)
    if isinstance(value, int) and not isinstance(value, bool):
        return str(value)
    if isinstance(value, float):
        return format(value, "g")
    if isinstance(value, list):
        return "[" + ",".join(canonicalize(item) for item in value) + "]"
    if isinstance(value, dict):
        return _canonicalize_object(value)
    raise TypeError(f"Unsupported JSON value type: {type(value)!r}")


def _canonicalize_object(value: dict[str, Any]) -> str:
    parts = []
    for key in sorted(value.keys()):
        parts.append(f"{escape_json_string(key)}:{canonicalize(value[key])}")
    return "{" + ",".join(parts) + "}"


def build_delivery_payload_json(request: dict[str, Any]) -> str:
    filtered = {key: value for key, value in request.items() if key in INCLUDED_FIELDS}
    return _canonicalize_object(filtered)


def compute_sha256_hex(json_value: Any) -> str:
    canonical_json = canonicalize(json_value)
    return hashlib.sha256(canonical_json.encode("utf-8")).hexdigest()


def compute_delivery_payload_sha256_hex(request: dict[str, Any]) -> str:
    delivery_json = build_delivery_payload_json(request)
    return compute_sha256_hex(json.loads(delivery_json))

File: examples/payload-hash/javascript/mail_payload_hash.mjs
import { createHash } from 'node:crypto';

export const INCLUDED_FIELDS = new Set([
  'source_service',
  'purpose',
  'to',
  'cc',
  'bcc',
  'subject',
  'html_body',
  'text_body',
  'reply_to',
  'metadata',
  'attachments',
]);

// attachments is included with a special projection (ADR 0022 D-03): absent or an empty array
// omits the field entirely from the hash document; a non-empty array is re-projected to exactly
// file_name (NFC), content_type, byte_length, content_sha256, and a zero-based order generated
// from array position -- never the raw content_base64 or an unverified declared content_type.
const ATTACHMENTS_FIELD_NAME = 'attachments';

// to/cc/bcc follow the same omission rule (ADR 0023 D-02): unspecified, null, and an empty
// array are all equivalent and the role is omitted from the hash document entirely (no
// "to":null or "to":[] ever appears; a CC-only request has no "to" key at all). A non-empty
// role is re-projected to the validated canonical recipient value (trimmed email;
// whitespace-only display_name treated as absent) -- not the raw request bytes -- so
// equivalent-but-differently-formatted submissions hash identically.
const RECIPIENT_FIELD_NAMES = ['to', 'cc', 'bcc'];

export function escapeJsonString(value) {
  let result = '"';
  for (const character of value) {
    switch (character) {
      case '"':
        result += '\\"';
        break;
      case '\\':
        result += '\\\\';
        break;
      case '\b':
        result += '\\b';
        break;
      case '\f':
        result += '\\f';
        break;
      case '\n':
        result += '\\n';
        break;
      case '\r':
        result += '\\r';
        break;
      case '\t':
        result += '\\t';
        break;
      default:
        if (character.charCodeAt(0) < 0x20) {
          result += `\\u${character.charCodeAt(0).toString(16).padStart(4, '0')}`;
        } else {
          result += character;
        }
        break;
    }
  }
  result += '"';
  return result;
}

export function canonicalize(value) {
  if (value === null) {
    return 'null';
  }
  if (typeof value === 'boolean') {
    return value ? 'true' : 'false';
  }
  if (typeof value === 'string') {
    return escapeJsonString(value);
  }
  if (typeof value === 'number') {
    if (Number.isInteger(value)) {
      return value.toString(10);
    }
    return value.toString();
  }
  if (Array.isArray(value)) {
    return `[${value.map((item) => canonicalize(item)).join(',')}]`;
  }
  if (typeof value === 'object') {
    const properties = Object.keys(value)
      .sort()
      .map((key) => `${escapeJsonString(key)}:${canonicalize(value[key])}`);
    return `{${properties.join(',')}}`;
  }
  throw new TypeError(`Unsupported JSON value type: ${typeof value}`);
}

function projectAttachments(attachments) {
  return attachments.map((attachment, order) => ({
    file_name: attachment.file_name.normalize('NFC'),
    content_type: attachment.content_type,
    byte_length: attachment.byte_length,
    content_sha256: attachment.content_sha256,
    order,
  }));
}

function projectRecipientRole(role) {
  if (!role || role.length === 0) {
    return null;
  }
  return role.map((recipient) => {
    const entry = { email: recipient.email.trim() };
    const displayName = recipient.display_name;
    if (displayName != null && displayName.trim() !== '') {
      entry.display_name = displayName;
    }
    return entry;
  });
}

export function buildDeliveryPayloadJson(request, attachments = null) {
  const filtered = {};
  for (const key of Object.keys(request)) {
    if (INCLUDED_FIELDS.has(key) && key !== ATTACHMENTS_FIELD_NAME && !RECIPIENT_FIELD_NAMES.includes(key)) {
      filtered[key] = request[key];
    }
  }
  for (const fieldName of RECIPIENT_FIELD_NAMES) {
    if (Object.prototype.hasOwnProperty.call(request, fieldName)) {
      const projected = projectRecipientRole(request[fieldName]);
      if (projected !== null) {
        filtered[fieldName] = projected;
      }
    }
  }
  if (attachments && attachments.length > 0) {
    filtered[ATTACHMENTS_FIELD_NAME] = projectAttachments(attachments);
  }

  const properties = Object.keys(filtered)
    .sort()
    .map((key) => `${escapeJsonString(key)}:${canonicalize(filtered[key])}`);
  return `{${properties.join(',')}}`;
}

export function computeSha256Hex(jsonValue) {
  const canonicalJson = canonicalize(jsonValue);
  return createHash('sha256').update(canonicalJson, 'utf8').digest('hex');
}

export function computeDeliveryPayloadSha256Hex(request, attachments = null) {
  const deliveryJson = buildDeliveryPayloadJson(request, attachments);
  return computeSha256Hex(JSON.parse(deliveryJson));
}
